### ISSUE TITLE
Fatal error:  Uncaught TypeError: ExpectationAnalyzer::isValidAtCall(): Argument #1 ($methodCall) must be of type PhpParser\Node\Expr\MethodCall

### ISSUE THREAD

**Issue by umpirsky** [rectorphp/rector#6976]
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | `^0.12.15` (`75c0b27d37ee34844b0eb5faf35742c6631d5ab2`)             |

Error:
```
./bin/rector                               
   11/2156 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%PHP Fatal error:  Uncaught TypeError: Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer::isValidAtCall(): Argument #1 ($methodCall) must be of type PhpParser\Node\Expr\MethodCall, PhpParser\Node\Expr\StaticCall given, called in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php on line 58 and defined in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php:83
Stack trace:
#0 vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php(58): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->isValidAtCall()
#1 vendor/rector/rector/vendor/rector/rector-phpunit/src/Rector/ClassMethod/MigrateAtToConsecutiveExpectationsRector.php(71): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->getExpectationsFromExpressions()
#2 vendor/rector/rector/src/Rector/AbstractRector.php(243): Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector->refactor()
#3 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176): Rector\Core\Rector\AbstractRector->enterNode()
#4 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#5 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#6 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#7 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#8 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85): PhpParser\NodeTraverser->traverseArray()
#9 vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(56): PhpParser\NodeTraverser->traverse()
#10 vendor/rector/rector/src/Application/FileProcessor.php(54): Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse()
#11 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(137): Rector\Core\Application\FileProcessor->refactor()
#12 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(101): Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors()
#13 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(175): Rector\Core\Application\FileProcessor\PhpFileProcessor->process()
#14 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(148): Rector\Core\Application\ApplicationFileProcessor->processFiles()
#15 vendor/rector/rector/src/Console/Command/ProcessCommand.php(147): Rector\Core\Application\ApplicationFileProcessor->run()
#16 vendor/rector/rector/vendor/symfony/console/Command/Command.php(282): Rector\Core\Console\Command\ProcessCommand->execute()
#17 vendor/rector/rector/vendor/symfony/console/Application.php(853): RectorPrefix20220126\Symfony\Component\Console\Command\Command->run()
#18 vendor/rector/rector/vendor/symfony/console/Application.php(289): RectorPrefix20220126\Symfony\Component\Console\Application->doRunCommand()
#19 vendor/rector/rector/src/Console/ConsoleApplication.php(67): RectorPrefix20220126\Symfony\Component\Console\Application->doRun()
#20 vendor/rector/rector/vendor/symfony/console/Application.php(187): Rector\Core\Console\ConsoleApplication->doRun()
#21 vendor/rector/rector/bin/rector.php(57): RectorPrefix20220126\Symfony\Component\Console\Application->run()
#22 vendor/rector/rector/bin/rector(5): require_once('...')
#23 {main}
  thrown in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php on line 83
Fatal error: Uncaught TypeError: Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer::isValidAtCall(): Argument #1 ($methodCall) must be of type PhpParser\Node\Expr\MethodCall, PhpParser\Node\Expr\StaticCall given, called in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php on line 58 and defined in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php:83
Stack trace:
#0 vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php(58): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->isValidAtCall()
#1 vendor/rector/rector/vendor/rector/rector-phpunit/src/Rector/ClassMethod/MigrateAtToConsecutiveExpectationsRector.php(71): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->getExpectationsFromExpressions()
#2 vendor/rector/rector/src/Rector/AbstractRector.php(243): Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector->refactor()
#3 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176): Rector\Core\Rector\AbstractRector->enterNode()
#4 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#5 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#6 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#7 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#8 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85): PhpParser\NodeTraverser->traverseArray()
#9 vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(56): PhpParser\NodeTraverser->traverse()
#10 vendor/rector/rector/src/Application/FileProcessor.php(54): Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse()
#11 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(137): Rector\Core\Application\FileProcessor->refactor()
#12 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(101): Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors()
#13 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(175): Rector\Core\Application\FileProcessor\PhpFileProcessor->process()
#14 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(148): Rector\Core\Application\ApplicationFileProcessor->processFiles()
#15 vendor/rector/rector/src/Console/Command/ProcessCommand.php(147): Rector\Core\Application\ApplicationFileProcessor->run()
#16 vendor/rector/rector/vendor/symfony/console/Command/Command.php(282): Rector\Core\Console\Command\ProcessCommand->execute()
#17 vendor/rector/rector/vendor/symfony/console/Application.php(853): RectorPrefix20220126\Symfony\Component\Console\Command\Command->run()
#18 vendor/rector/rector/vendor/symfony/console/Application.php(289): RectorPrefix20220126\Symfony\Component\Console\Application->doRunCommand()
#19 vendor/rector/rector/src/Console/ConsoleApplication.php(67): RectorPrefix20220126\Symfony\Component\Console\Application->doRun()
#20 vendor/rector/rector/vendor/symfony/console/Application.php(187): Rector\Core\Console\ConsoleApplication->doRun()
#21 vendor/rector/rector/bin/rector.php(57): RectorPrefix20220126\Symfony\Component\Console\Application->run()
#22 vendor/rector/rector/bin/rector(5): require_once('...')
```

`rector.php`

```
<?php

declare(strict_types=1);

use Rector\Core\Configuration\Option;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    // get parameters
    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [
        __DIR__ . '/tests'
    ]);

    // Define what rule sets will be applied
    // $containerConfigurator->import(LevelSetList::UP_TO_PHP_74);

    // get services (needed for register a single rule)
    $services = $containerConfigurator->services();

    // register a single rule
    $services->set(Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector::class);
};
```

**Comment by TomasVotruba** [comment]
Thank you for your report!

We'll need an isolated failing demo link from: http://getrector.org/demo,
that way we can reproduce the bug.

**Comment by umpirsky** [comment]
Unfortunately I can't find the code that causes the issue. Is there a way to know what test causes it?

**Comment by TomasVotruba** [comment]
Try running with `--debug` option, it will show more details.

**Comment by umpirsky** [comment]
Found it, but can't reproduce on https://getrector.org/demo

It breaks on last line:

```
$a = $this->createMock(A::class);
$a->expects(self::once())->method('getId')->willReturn(1);
```

**Comment by TomasVotruba** [comment]
Looks like PHPUnit code. That would require full PHPUnit dependency.
Could you make a reproducible repository on GitHub?

**Comment by umpirsky** [comment]
Added on https://github.com/umpirsky/rector-phpunit.

But looks like I can get it reproduced only once. Same in my project when I work on single file.

If you want to reproduce again, just rename the file (or clear the cache if there is one).

**Comment by TomasVotruba** [comment]
Thanks :+1: , I'm looking into it

**Comment by TomasVotruba** [comment]
Fixed in https://github.com/rectorphp/rector-phpunit/pull/41

**Comment by umpirsky** [comment]
Thanks!

**Comment by umpirsky** [comment]
It works. :+1:

**Comment by TomasVotruba** [comment]
Glad to help :+1: Thanks for the minimalistic repository, it made the fix quite easy to find

**Comment by umpirsky** [comment]
I get another one on:

```
use GuzzleHttp\Handler\CurlHandler;
use PHPUnit\Framework\TestCase;

class HttpFallbackMiddlewareTest extends TestCase
{
    public function testFoo($promise, $expection)
    {
        $handler = $this->createMock(CurlHandler::class);
        $handler
            ->expects($expection)
            ->method('__invoke')
            ->willReturn($promise);
    }
}
```

```
./bin/rector --debug
[parsing] tests/CoreBundle/PlatformService/HttpClient/HttpFallbackMiddlewareTest.php
[refactoring] tests/CoreBundle/PlatformService/HttpClient/HttpFallbackMiddlewareTest.php
    [applying] Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector
PHP Fatal error:  Uncaught TypeError: Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer::isValidAtCall(): Argument #1 ($call) must be of type PhpParser\Node\Expr\MethodCall|PhpParser\Node\Expr\StaticCall, PhpParser\Node\Expr\Variable given, called in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php on line 64 and defined in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php:102
Stack trace:
#0 vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php(64): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->isValidAtCall()
#1 vendor/rector/rector/vendor/rector/rector-phpunit/src/Rector/ClassMethod/MigrateAtToConsecutiveExpectationsRector.php(76): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->getExpectationsFromExpressions()
#2 vendor/rector/rector/src/Rector/AbstractRector.php(243): Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector->refactor()
#3 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176): Rector\Core\Rector\AbstractRector->enterNode()
#4 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#5 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#6 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#7 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#8 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85): PhpParser\NodeTraverser->traverseArray()
#9 vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(56): PhpParser\NodeTraverser->traverse()
#10 vendor/rector/rector/src/Application/FileProcessor.php(54): Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse()
#11 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(137): Rector\Core\Application\FileProcessor->refactor()
#12 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(101): Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors()
#13 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(175): Rector\Core\Application\FileProcessor\PhpFileProcessor->process()
#14 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(148): Rector\Core\Application\ApplicationFileProcessor->processFiles()
#15 vendor/rector/rector/src/Console/Command/ProcessCommand.php(147): Rector\Core\Application\ApplicationFileProcessor->run()
#16 vendor/rector/rector/vendor/symfony/console/Command/Command.php(282): Rector\Core\Console\Command\ProcessCommand->execute()
#17 vendor/rector/rector/vendor/symfony/console/Application.php(853): RectorPrefix20220126\Symfony\Component\Console\Command\Command->run()
#18 vendor/rector/rector/vendor/symfony/console/Application.php(289): RectorPrefix20220126\Symfony\Component\Console\Application->doRunCommand()
#19 vendor/rector/rector/src/Console/ConsoleApplication.php(67): RectorPrefix20220126\Symfony\Component\Console\Application->doRun()
#20 vendor/rector/rector/vendor/symfony/console/Application.php(187): Rector\Core\Console\ConsoleApplication->doRun()
#21 vendor/rector/rector/bin/rector.php(57): RectorPrefix20220126\Symfony\Component\Console\Application->run()
#22 vendor/rector/rector/bin/rector(5): require_once('...')
#23 {main}
  thrown in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php on line 102
Fatal error: Uncaught TypeError: Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer::isValidAtCall(): Argument #1 ($call) must be of type PhpParser\Node\Expr\MethodCall|PhpParser\Node\Expr\StaticCall, PhpParser\Node\Expr\Variable given, called in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php on line 64 and defined in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php:102
Stack trace:
#0 vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php(64): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->isValidAtCall()
#1 vendor/rector/rector/vendor/rector/rector-phpunit/src/Rector/ClassMethod/MigrateAtToConsecutiveExpectationsRector.php(76): Rector\PHPUnit\NodeAnalyzer\ExpectationAnalyzer->getExpectationsFromExpressions()
#2 vendor/rector/rector/src/Rector/AbstractRector.php(243): Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector->refactor()
#3 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176): Rector\Core\Rector\AbstractRector->enterNode()
#4 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#5 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#6 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105): PhpParser\NodeTraverser->traverseArray()
#7 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196): PhpParser\NodeTraverser->traverseNode()
#8 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85): PhpParser\NodeTraverser->traverseArray()
#9 vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(56): PhpParser\NodeTraverser->traverse()
#10 vendor/rector/rector/src/Application/FileProcessor.php(54): Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse()
#11 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(137): Rector\Core\Application\FileProcessor->refactor()
#12 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(101): Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors()
#13 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(175): Rector\Core\Application\FileProcessor\PhpFileProcessor->process()
#14 vendor/rector/rector/src/Application/ApplicationFileProcessor.php(148): Rector\Core\Application\ApplicationFileProcessor->processFiles()
#15 vendor/rector/rector/src/Console/Command/ProcessCommand.php(147): Rector\Core\Application\ApplicationFileProcessor->run()
#16 vendor/rector/rector/vendor/symfony/console/Command/Command.php(282): Rector\Core\Console\Command\ProcessCommand->execute()
#17 vendor/rector/rector/vendor/symfony/console/Application.php(853): RectorPrefix20220126\Symfony\Component\Console\Command\Command->run()
#18 vendor/rector/rector/vendor/symfony/console/Application.php(289): RectorPrefix20220126\Symfony\Component\Console\Application->doRunCommand()
#19 vendor/rector/rector/src/Console/ConsoleApplication.php(67): RectorPrefix20220126\Symfony\Component\Console\Application->doRun()
#20 vendor/rector/rector/vendor/symfony/console/Application.php(187): Rector\Core\Console\ConsoleApplication->doRun()
#21 vendor/rector/rector/bin/rector.php(57): RectorPrefix20220126\Symfony\Component\Console\Application->run()
#22 vendor/rector/rector/bin/rector(5): require_once('...')
#23 {main}
  thrown in vendor/rector/rector/vendor/rector/rector-phpunit/src/NodeAnalyzer/ExpectationAnalyzer.php on line 102
```

**Comment by TomasVotruba** [comment]
I'll check it... possibly string too :)

**Comment by TomasVotruba** [comment]
Next PR fix: https://github.com/rectorphp/rector-phpunit/pull/42

**Comment by umpirsky** [comment]
That worked, thanks! :+1:

**Comment by umpirsky** [comment]
@TomasVotruba Added new example in Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector

```php
<?php

namespace Tests\AdminBundle\Form\DataTransformer\Drill;

use PHPUnit\Framework\TestCase;

class ProblematicTest extends TestCase
{
    private $foo;

    public function barTest()
    {
        $i = 0;

        $this->foo
            ->expects($this->at($i))
            ->method('createFromSeeks')
            ->willReturn(1);
    }
}
```


```php
<?php

use Rector\Php74\Rector\Property\TypedPropertyRector;
use Rector\Set\ValueObject\SetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $parameters = $containerConfigurator->parameters();
    $services = $containerConfigurator->services();

    // A. run whole set
    $containerConfigurator->import(SetList::DEAD_CODE);

    // B. or single rule
    $services->set(Rector\PHPUnit\Rector\ClassMethod\MigrateAtToConsecutiveExpectationsRector::class);
};
```

For some reason it didn't fix `at()`. Should I open a new issue?

**Comment by umpirsky** [comment]
Another example https://getrector.org/demo/1ec84192-9d46-65fc-947d-9f7ed7e29daf

**Comment by samsonasik** [comment]
@umpirsky demo page is not using latest "rector/rector": "dev-main", it using tagged version, please try using latest dev-main

```bash
composer config minimum-stability dev
composer config prefer-stable true
composer update rector/rector:dev-main
```

**Comment by TomasVotruba** [comment]
> Should I open a new issue?

We'll need a test fixture in the rule itself. See my last 2 PRs on how to add them.

**Comment by TomasVotruba** [comment]
@umpirsky After few hours of digging, I think this case is way too ambiguous to handle with Rector. It should be done manualy.
See https://github.com/rectorphp/rector-phpunit/pull/44
